### ISSUE TITLE
Add support for EIP-1559 transaction pricing

### ISSUE THREAD

**Issue by clamdad** [tellor-io/telliot-core#166]
The current `contract.write()` method uses the deprecated `gasPrice` parameter, which sometimes causes higher transaction costs than are necessary.  It should be updated to use the new `maxFeePerGas` and `maxPriorityFeePerGas` interface for EIP-1559 transactions.

**Comment by tallywiesenberg** [comment]
I agree! at least for profit optimizations

**Comment by clamdad** [comment]
Contract.write should be updated to take both `maxFeePerGas` and `maxPriorityFeePerGas`.

`maxPriorityFeePerGas`: optional - default: use value from call `Eth.max_priority_fee`.
`maxFeePerGas`: Optional, default: None (don't use).

**Comment by clamdad** [comment]
Question: Should we maintain support for current gas price method?  Some chains might not support EIP-1119 yet.

**Comment by tallywiesenberg** [comment]
I think we should. I definitely think other chains are lagging on eip-1559 implementation...polygon is just now implementing it https://blog.polygon.technology/polygon-is-bringing-eip-1559-upgrades-to-its-network-with-a-testnet-release/

**Comment by clamdad** [comment]
Sounds like a good approach.  There's quite a bit of potential for confusion here (I'm still a little fuzzy), so documentation of the `Contract.write` API should be really solid.  

Maybe we should even take a stab at that first.  Can you summarize how it would work in `Contract.write` comment string, using actual variable names, and (clarifying when legacy and eip-1559 are used).   

We probably also want consistent CLI flags to handle these gas-related items.

At the next higher layer, we also want to be able to choose a gas price source, such as the new `telliot_core.gas.etherscan_gas.EtherscanGasPriceSource` source.

**Comment by tallywiesenberg** [comment]
> Sounds like a good approach. There's quite a bit of potential for confusion here (I'm still a little fuzzy), so documentation of the `Contract.write` API should be really solid.
> 
> Maybe we should even take a stab at that first. Can you summarize how it would work in `Contract.write` comment string, using actual variable names, and (clarifying when legacy and eip-1559 are used).
> 
> We probably also want consistent CLI flags to handle these gas-related items.
> 
> At the next higher layer, we also want to be able to choose a gas price source, such as the new `telliot_core.gas.etherscan_gas.EtherscanGasPriceSource` source.

I agree. The bullet points above are exactly how i implemented it in `Contract.write`:

```python
#start tx dict with static elements
            tx_dict = {
                "from": acc.address,
                "nonce": acc_nonce,
                "gas": gas_limit,
                "chainId": self.node.chain_id
            }

            #if legacy gas price is not None, use it
            if gas_price is not None:
                tx_dict["gasPrice"] = self.node.web3.toWei(gas_price, "gwei")

            #else if (if legacy gas price is None) and max fee is not None, use max fee
            elif max_fee_per_gas is not None:
                tx_dict["maxFeePerGas"] = self.node.web3.toWei(max_fee_per_gas, "gwei")

            #else if (if legacy price and max fee are not provided), use max priority fee
            elif max_priority_fee_per_gas is not None:
                tx_dict["maxPriorityFeePerGas"] = self.node.web3.toWei(max_priority_fee_per_gas, "gwei")

            #else (if none are provided) estimate max priority fee with rpc node
            else:
                tx_dict["maxPriorityFeePerGas"] = self.node.web3.eth.max_priority_fee
            #pass in tx dict to build the transaction
            built_tx = transaction.buildTransaction(tx_dict)
            # submit transaction
            tx_signed = acc.sign_transaction(built_tx)

```

**Comment by tallywiesenberg** [comment]
let me know if you have any further questions @clamdad

**Comment by clamdad** [comment]
Is it possible to use both `max_fee_per_gas` and `max_priority_fee_per_gas`?

**Comment by tallywiesenberg** [comment]
> We probably want to allow both `max_fee_per_gas` and `max_priority_fee_per_gas`, since the standard allows it.

good point, i shall work this in!

**Comment by tallywiesenberg** [comment]
now factored into PR ✅
